### ISSUE TITLE
Download Method

### ISSUE THREAD

**Issue by Matt-Gleich** [Standard-Structure/struct#1]
Figure out a way to download the structures directly from GitHub or using some caching method.

**Comment by Matt-Gleich** [comment]
https://stackoverflow.com/questions/33066582/how-to-download-a-folder-from-github/48948711

**Comment by Cal-Hagner** [comment]
@rishiosaur I know you want to expose CLI functionality that can be used directly or via the CLI tool. I've never used gluegun before, is that something an existing directory already exists for (extensions?), or should we be creating a new directory and then import from there?

**Comment by rishiosaur** [comment]
Creating an `index.ts` to export everything from seems like the best way to go. I guess we would just make another directory called ‘struct’ in the src/ folder.

**Comment by Cal-Hagner** [comment]
> https://stackoverflow.com/questions/33066582/how-to-download-a-folder-from-github/48948711

We're going to want to add `-C <path-to-cache-repo>` to the `tar` part of his command so that struct puts the downloaded the folder in the cache and then creates a new project based on that cached folder (downloading for the cache based on our requirements: have to be online and the diffs between the cache and Standard-Structure have to reflect changes in Standard-Structure.

**Comment by cjdenio** [comment]
I don't think the `tar` command works on Windows. We may need to find some library for Node that does un-`tar`-ing for us. This has probably been thought through, but I'm putting it here just in case :smile: 

Also, will this require `struct` to download the entire `Standard-Structure` repo when someone generates a project structure?

**Comment by Matt-Gleich** [comment]
This npm package should work for you guys if you are just trying to extract the files from the downloaded zip:
https://www.npmjs.com/package/zip

**Comment by cjdenio** [comment]
We could probably find something more modern, but yeah, that would help :grin:

**Comment by Matt-Gleich** [comment]
Actually, why don't we just clone the repo when the user first uses the cli and run git pull if the user has a network connection? This could also serve as the caching mechanism. We could use [lib2git](https://libgit2.org/) to run the git commands even if the user doesn't have git.

**Comment by Matt-Gleich** [comment]
More specifically [nodegit](https://github.com/nodegit/nodegit)

**Comment by cjdenio** [comment]
That would take care of the issue we discussed yesterday where the GitHub repo might be different from the local cache. I don't think that requiring a Git installation is too much of a stretch, though, because most users will already have it.

**Comment by Cal-Hagner** [comment]
That would work really well.

**Comment by Cal-Hagner** [comment]
Should we assume that users already have Git?

If we do that, we can use sparse checkout to only clone directories on a user-need basis. Otherwise, I foresee a potential issue (once Standard-Structure has more packages) where cloning the repository initially becomes slow and the clone takes up a lot of unnecessary space on the user's file system. What do you think? @cjdenio @Matt-Gleich

**Comment by Matt-Gleich** [comment]
We could have it so that in the future you can run a command such as `struct trim` that removes all structs you haven't used before. This means that whenever you make a structure it registers a use.

**Comment by Cal-Hagner** [comment]
What if someone tries to create a structure that they just removed with `struct trim`? We still can't selectively download that needed struct, so we need to reclone everything (or just pull), downloading everything, which brings us back to square 1.

**Comment by Cal-Hagner** [comment]
I have a feeling that there is a more elegant solution here, and requiring Git may be one way. We should also poke around in the GitHub API for ways to check if new commits have been added.

**Comment by cjdenio** [comment]
We _could_ just create separate repos for different structures. Not pretty, but functional. Then, we could setup some Git submodules to pull those repos into `Standard-Structure`.

**Comment by Matt-Gleich** [comment]
Oh yeah that can definitely be done with the GitHub API. I think that makes the most sense here.

**Comment by Cal-Hagner** [comment]
@cjdenio I'm game for that, but where do we draw the line? Do we use submodules for each top-level framework/language, or submodules for every variation on those languages and frameworks? Either way, I pinned the Standard-Structure and struct repos so that we'll never lose sight of them even if we have to develop separate repos for everything.

**Comment by Matt-Gleich** [comment]
That is going to be tremendously hard to manage. We could be looking at 100 repos that our small core team has to manage. Using the GitHub API is a much more elegant solution and is WAY more scalable.

**Comment by cjdenio** [comment]
I completely agree with @Matt-Gleich. I didn't think we'd be this big when I suggested it :joy:

**Comment by Cal-Hagner** [comment]
Okay. I'm a big fan of going the GitHub API way myself.

**Comment by cjdenio** [comment]
OK; I did some research, and it looks like we have a couple downloading options that use the GitHub API.

First, we could use the [Contents API](https://developer.github.com/v3/repos/contents/) to list all files/directories, then recursively download every file in the structure.

**Pros:**
- Doesn't download the whole repo

**Cons:**
- Uses a whole bunch of API requests, so we could quickly run into rate limits.
- Super hard to do

Second, we could use the Contents API (again) to download a .zip (or .tar.gz) of the entire repo, then extract what we need (or keep the whole thing)

**Pros:**
- Only takes up 1 API request.
- Super easy
- Makes the entire (compressed) structure library available on the user's computer (could be a good thing, could be a bad thing)

**Cons:**
- Requires an unzipping library
- Downloads the entire repo, so it could take longer to download as the project grows.

**Comment by Matt-Gleich** [comment]
I think option two makes the most sense. We don't wanna be running into rate limiting issues. We can also just download a zip of one folder and not the whole repo as seen in:

https://stackoverflow.com/questions/33066582/how-to-download-a-folder-from-github/48948711

**Comment by cjdenio** [comment]
I'm reading the SO post. Which answer are you looking at?

**Comment by Matt-Gleich** [comment]
The one showing the example command of:
```
curl https://codeload.github.com/zeit/next.js/tar.gz/master | \
  tar -xz --strip=2 next.js-master/examples/with-apollo
```

**Comment by cjdenio** [comment]
If we can find a cross-platform way of doing that, I'd say that's our best option.

**Comment by Cal-Hagner** [comment]
This brings us back to square 1 in terms of keeping the local cached directories up to date with changes in the GitHub repo. How can we check the diffs between the cached directories and master?

The Contents API is Rest v3, I've been playing around with the v4 GraphQL API, and I'm going to play with it some more, but I definitely think there's a way we can selectively clone and then check the commit history for commits that change files that the user has cached. It may require storing metadata on the user's cache inside the cache directory, but only downloading on a user-need basis seems to me to be the cleanest way of doing things, and I don't see the 60 requests per hour being a huge limit.

**Comment by cjdenio** [comment]
I was thinking that we could store the hash of the latest local commit in a config file, then re-download if the latest remote commit's hash differs.

**Comment by Matt-Gleich** [comment]
That would only let us know if the repo updated, not if the used structures have updated

**Comment by Cal-Hagner** [comment]
I like the commit hashing idea. This a GraphQL query that is close to what we want:

```
{
  organization(login: "Standard-Structure") {
    repository(name: "Standard-Structure") {
      object(expression: "712d5f2") {
        ... on Commit {
          message
        }
      }
    }
  }
}
```

Sadly, the GitHub v4 API currently doesn't support changed files. So, I went poking around the v3 API to see what I could find. Turns out that by making only one request, we can compare two commits by SHA. This does require an additional request to get the latest commit's SHA from the repo, but it also gives us an object called `files` that contains all the files changed between the two commits. You can [see for yourself at the v3 Commit API documentation](https://developer.github.com/v3/repos/commits/#compare-two-commits). From there, we can figure out a way to get the new files, but this gives us a way to store the SHA metadata in the cache and then update whenever the user creates a new project using one of our structures.

Hopefully the GraphQL API will move in a direction that supports this soon, so that we can then make a single request and potentially get everything we need. Ah, if wishes were horses...

Anyway, I think this may be a good solution. Any thoughts? @cjdenio @Matt-Gleich

**Comment by rishiosaur** [comment]
Sorry I haven't been super active on this repo lately, I've been organizing a lot on the other repo. I really like the idea of hashing commits; it's a lot easier to compare than a full recursive search of the cached directories. My only concern is about storing API Keys. Should we just have our own obfuscation somewhere or have the users generate their own tokens?

**Comment by rishiosaur** [comment]
Obviously, we could also have some form of auth functionality

**Comment by Cal-Hagner** [comment]
That's a valid concern.

**Comment by Matt-Gleich** [comment]
It should be fine. Just bring up an OAuth page like any other GitHub application. This article might help with the actual storing of them:
https://www.freecodecamp.org/news/how-to-securely-store-api-keys-4ff3ea19ebda/

**Comment by cjdenio** [comment]
Just want to put this up here for discussion: how often will an individual structure be updated? I imagine that it would only happen if a bug was found, which isn't likely with a bare-bones project structure.

So, is it even worth it to build in an auto-update? We'd obviously have a command like `struct cache refresh` to re-download cached structures, but do we really need to automatically re-download them?

Just a thought.

**Comment by Matt-Gleich** [comment]
I think maybe something we should talk about during our meeting this week.

**Comment by cjdenio** [comment]
Now that `struct` can download structures, can this be closed?

**Comment by Matt-Gleich** [comment]
I think so, worse case scenario we just reopen the issue.

**Comment by cjdenio** [comment]
Wow, we just closed #1 :exploding_head:
